test/test_remote_io.py

```diff
@@ -51,6 +54,46 @@ def test_http_reader_iterdatapipe(self):
         http_dp = HttpReader(source_dp)
         self.assertEqual(1, len(http_dp))
 
+    def test_on_disk_cache_holder_iterdatapipe(self):
+        file_url = "https://raw.githubusercontent.com/pytorch/data/main/LICENSE"
+        prefix = "OnDisk_"
+        expected_file_name = os.path.join(self.temp_dir.name, prefix + "LICENSE")
+        expected_MD5_hash = "4aabe940637d4389eca42ac1a0e874ec"
+
+        file_dp = IterableWrapper([file_url])
+
+        def _filepath_fn(url):
+            filename = prefix + os.path.basename(url)
+            return os.path.join(self.temp_dir.name, filename)
+
+        cache_dp = file_dp.on_disk_cache(filepath_fn=_filepath_fn, extra_check_fn=partial(check_hash_fn, expected_hash=expected_MD5_hash, hash_type="md5"))
+
+        # DataPipe Constructor
+        cache_dp = HttpReader(cache_dp)
+        # Functional API
+        cache_dp = cache_dp.map(fn=lambda x: b''.join(x), input_col=1)
+
+        # Start iteration without `end_caching`
+        with self.assertRaisesRegex(RuntimeError, "Please call"):
+            _ = list(cache_dp)
+
+        cache_dp = cache_dp.end_caching(mode="wb")
+
+        # File doesn't exist on disk
+        self.assertFalse(os.path.exists(expected_file_name))
+        path = list(cache_dp)[0]
+        self.assertTrue(os.path.exists(expected_file_name))
+
+        # File is cached to disk
+        self.assertEqual(expected_file_name, path)
+        self.assertTrue(check_hash_fn(path, expected_MD5_hash, hash_type="md5"))
+
+        # Call `end_caching` again
+        with self.assertRaisesRegex(RuntimeError, "Incomplete `OnDiskCacheHolder` is required in the pipeline"):
+            cache_dp = cache_dp.end_caching()
+
+        # TODO(ejguan): Multiple CacheHolders or nested CacheHolders
```

Comment:
I will add test for multiple cachehodlers in sequence if the API looks fine.
```py
cache_dp1 = dp.on_disk_cache().op1().op2().end_caching()
cache_dp2 = cache_dp1.on_disk_cache().op3().op4().end_caching()
```
And, as discussed with Vitaly, we are probably not going to allow nested `cachehodler`.
```py
cache_dp = dp.on_disk_cache().op1().op2().on_disk_cache().op3().op4().end_caching().end_caching()
```